daemon/cluster/cluster.go

```diff
@@ -871,6 +872,36 @@ func (c *Cluster) GetServices(options apitypes.ServiceListOptions) ([]types.Serv
 	return services, nil
 }
 
+// imageWithDigestString takes an image such as name or name:tag
+// and returns the image pinned to a digest, such as name@sha256:34234...
+func (c *Cluster) imageWithDigestString(ctx context.Context, image, encodedAuth string) (string, error) {
+	// TODO(nishanttotla): create AuthConfig from encodedAuth
+
+	ref, err := reference.ParseNamed(image)
+	if err != nil {
+		return "", err
+	}
+	// only pin if not a canonical reference (i.e. with digest)
+	if _, ok := ref.(reference.Canonical); !ok {
+		ref = reference.WithDefaultTag(ref)
+
+		namedTaggedRef, ok := ref.(reference.NamedTagged)
+		if !ok {
+			return "", fmt.Errorf("unable to cast image to NamedTagged reference object")
+		}
+		dgst, err := c.config.Backend.ResolveTagToDigest(ctx, namedTaggedRef, &apitypes.AuthConfig{})
+		if err != nil {
+			return "", err
+		}
+		refDgst, err := reference.WithDigest(ref, dgst)
+		if err != nil {
+			return "", err
+		}
+		return refDgst.String(), nil
+	}
+	return "", fmt.Errorf("error while trying to retrieve image digest")
```

Comment:
Don't make the error path in the main path. Get in the habit of making an early return.